lib/sudo-exec/src/backchannel.rs

```diff
@@ -0,0 +1,158 @@
+use std::{
+    ffi::c_int,
+    io::{self, Read, Write},
+    mem::size_of,
+    os::unix::{net::UnixStream, process::ExitStatusExt},
+    process::ExitStatus,
+};
+
+use sudo_system::interface::ProcessId;
+
+type Prefix = u8;
+type ParentData = c_int;
+const PREFIX_LEN: usize = size_of::<Prefix>();
+const PARENT_DATA_LEN: usize = size_of::<ParentData>();
+
+pub(crate) fn backchannel_pair() -> io::Result<(ParentBackchannel, MonitorBackchannel)> {
+    let (sock1, sock2) = UnixStream::pair()?;
+    sock1.set_nonblocking(true)?;
+    sock2.set_nonblocking(true)?;
+    Ok((
+        ParentBackchannel { socket: sock1 },
+        MonitorBackchannel { socket: sock2 },
+    ))
+}
+
+pub(crate) enum ParentEvent {
+    IoError(c_int),
+    CommandExit(c_int),
+    CommandSignal(c_int),
+    CommandPid(ProcessId),
+}
+
+impl ParentEvent {
+    const LEN: usize = PREFIX_LEN + PARENT_DATA_LEN;
+}
+
+impl<'a> From<&'a io::Error> for ParentEvent {
+    fn from(err: &'a io::Error) -> Self {
+        // This only panics if an error is created using `io::Error::new`.
+        Self::IoError(err.raw_os_error().unwrap())
+    }
+}
+
+impl From<ExitStatus> for ParentEvent {
+    fn from(status: ExitStatus) -> Self {
+        if let Some(code) = status.code() {
+            Self::CommandExit(code)
+        } else {
+            // `ExitStatus::code` docs state that it only returns `None` if the process was
+            // terminated by a signal so this should always succeed.
+            Self::CommandSignal(status.signal().unwrap())
+        }
+    }
+}
+
+/// A socket use for commmunication between the monitor and the parent process.
+pub(crate) struct ParentBackchannel {
+    socket: UnixStream,
+}
+
+impl ParentBackchannel {
+    /// Send a [`MonitorEvent`].
+    ///
+    /// Calling this method will block until the socket is ready for writing.
+    pub(crate) fn send(&mut self, event: &MonitorEvent) -> io::Result<()> {
+        let prefix: Prefix = match event {
+            MonitorEvent::ExecCommand => 0,
+        };
+
+        let buf: [u8; MonitorEvent::LEN] = prefix.to_ne_bytes();
+
+        self.socket.write_all(&buf)
+    }
+
+    /// Receive a [`ParentEvent`].
+    ///
+    /// Calling this method will block until the socket is ready for reading.
+    pub(crate) fn recv(&mut self) -> io::Result<ParentEvent> {
+        let mut buf = [0; ParentEvent::LEN];
+        self.socket.read_exact(&mut buf)?;
+
+        let (prefix_buf, data_buf) = buf.split_at(PREFIX_LEN);
+
+        let prefix = Prefix::from_ne_bytes(prefix_buf.try_into().unwrap());
+        let data = ParentData::from_ne_bytes(data_buf.try_into().unwrap());
+
+        let event = match prefix {
+            0 => ParentEvent::IoError(data),
```

Comment:
could you please refactor these magic values into constants for readability? I think it would be less error prone that way.

It'd be nice to have the serialization and deserialization as methods on the Event type so maybe something like this

``` rust
impl ParentEvent {
    const IO_ERROR: Prefix = 0;
    const COMMAND_EXIT: Prefix = 1;
    // ..

    fn from_parts(prefix: prefix, data: ParentData) -> Self { /* .. */ }
    fn to_parts(&self) -> (Prefix, ParentData) { /* .. */ }
}
```